mesonbuild/interpreter/mesonmain.py

```diff
@@ -277,23 +280,30 @@ def install_dependency_manifest_method(self, args, kwargs):
         self.build.dep_manifest_name = args[0]
 
     @FeatureNew('meson.override_find_program', '0.46.0')
-    @permittedKwargs({})
-    def override_find_program_method(self, args, kwargs):
-        if len(args) != 2:
-            raise InterpreterException('Override needs two arguments')
-        name, exe = args
-        if not isinstance(name, str):
-            raise InterpreterException('First argument must be a string')
-        exe = unholder(exe)
+    @FeatureNewKwargs('meson.override_find_programs', '0.58.0', ['native'])
+    @permittedKwargs({'native'})
+    @typed_pos_args('meson.override_find_programs', str, (ExecutableHolder, mesonlib.File, ExternalProgramHolder))
+    def override_find_program_method(self, args: T.Tuple[str, T.Union[ExecutableHolder, mesonlib.File, ExternalProgramHolder]], kwargs: T.Dict[str, T.Any]):
+        name = args[0]
+        exe: T.Union[build.Executable, mesonlib.File, ExternalProgram] = unholder(args[1])
+
+        # We need native when using a file, and someone might want to override
+        # something for a specific machine, otherwise use the for_machien from
+        # the program itself.
+        if isinstance(exe, mesonlib.File) or 'native' in kwargs:
+            for_machine = self.interpreter.machine_from_native_kwarg(kwargs)
+        else:
+            for_machine = exe.for_machine
+
         if isinstance(exe, mesonlib.File):
             abspath = exe.absolute_path(self.interpreter.environment.source_dir,
                                         self.interpreter.environment.build_dir)
             if not os.path.exists(abspath):
                 raise InterpreterException('Tried to override %s with a file that does not exist.' % name)
-            exe = OverrideProgram(name, abspath)
+            exe = OverrideProgram(name, abspath, for_machine=for_machine)
         if not isinstance(exe, (ExternalProgram, build.Executable)):
             raise InterpreterException('Second argument must be an external program or executable.')
```

Comment:
nitpicking: this check is now useless, it's guaranteed to be one those 2 types. Removing those is the point of typed_pos_args.